### PR TITLE
Restructuring our OpenVR plugin

### PR DESCRIPTION
**Pull request by BastiaanOlij**
I'm far from finished here, but I'm restructuring our plugin to use godot-cpp. 

This will likely become version 2.0 of our plugin, one of the things I hope to change here is to expose the controller mesh through the new mesh functionality and handle this without pausing our thread but instead implementing proper thread handling for our controller.

I'm also slowly rewriting code to proper classes.

### CODE REVIEW

**Comment by BastiaanOlij** [comment]
Made good progress today to get the main plugin to compile and the openvr_data structure rewritten to a singleton class. I've also merged the configuration class that Benedikt made into this. There is a little more to do around here because I want to grab the singleton as soon as the interface instantiates, not wait until we initialise. That should ensure this object is available from start to finish and we don't loose config.

The render model and overlay objects I renamed to cpp_ just so they aren't compiled. I'm going to be rewriting these to use godot-cpp classes.

**Comment by BastiaanOlij** [comment]
Making some good progress with this. Controller mesh now loads async!! Still need to do textures and get ben's classes back up and running

File: SConstruct [hunk 1]
```diff
@@ -1,93 +1,134 @@
...
+env.Append(LIBS=[godot_cpp_library])
+
+# Add our openvr library
+platform_dir += str(env['bits'])
 
 if (os.name == "nt" and os.getenv("VCINSTALLDIR")):
-    env.Append(LIBPATH=[openvr_path + 'lib/' + platform_dir])
+    env.Append(LIBPATH=[env['openvr_path'] + 'lib/' + platform_dir])
     env.Append(LINKFLAGS=['openvr_api.lib'])
 elif platform == "osx":
```
Comment (beniwtv) [inline SConstruct]: Got an Scons error here. I think this line should be

`elif env['platform'] == "osx":`

Comment (BastiaanOlij) [reply SConstruct]: I think so too :)

**Comment by beniwtv** [comment]
Seems all good with the overlay! :)

I noticed the demo spams:
Condition ' p_tracker_id == 0 ' is true. returned: __null

It's because the controller ID for the left hand is set to 0 -> could be set to 1 (right hand is set to 2)

**Comment by BastiaanOlij** [comment]
> 
> 
> Seems all good with the overlay! :)
> 
> I noticed the demo spams:
> Condition ' p_tracker_id == 0 ' is true. returned: __null
> 
> It's because the controller ID for the left hand is set to 0 -> could be set to 1 (right hand is set to 2)

I'll have to remember to change the defaults in the demo app

**Comment by BastiaanOlij** [comment]
Ok, texture async load should work now as well. It could result in the controller showing without texture for a split second but this should solve the hickup we've always had when turning a controller off (assuming the shader cache is used to precompile the shader).

ARVRController.cpp could use some cleanup I guess but I'm not against it not using the CPP classes. 

With this stuff done we've got a good base to build upon. Next step is looking into improving how we handle meshes and start looking into gestures but those will be new PRs :)

Please test so we can merge:)

**Comment by beniwtv** [comment]
So I started testing with 3.1.1 stable and it crashes when setting the controller texture.
Compiling Godot master and it works however.

I see spam in the logs over and over again, but it is actually a Godot bug I think:

`ERROR: recover_from_penetration: Condition ' shape_idx < 0 || shape_idx >= cs->getNumChildShapes() ' is true. returned: false
   At: modules/bullet/space_bullet.cpp:1234.
`

**Comment by BastiaanOlij** [comment]
yeah, that one is a known one, I think it was fixed, are you on a recent master?

**Comment by beniwtv** [comment]
Yes, and I've not since seen this error again, so all fine :)

**Comment by BastiaanOlij** [comment]
Time to merge you think?

**Comment by beniwtv** [comment]
I think it's good to go :)

**Pull request merged**